libbpf-tools/javagc.c

```diff
@@ -0,0 +1,237 @@
+{
+	if (level == LIBBPF_DEBUG && ! env.verbose)
+		return 0;
+
+	return vfprintf(stderr, format, args);
+}
+
+static void bump_memlock_rlimit(void)
+{
+	struct rlimit rlim_new = {
+		.rlim_cur	= RLIM_INFINITY,
+		.rlim_max	= RLIM_INFINITY,
+	};
+
+	if (setrlimit(RLIMIT_MEMLOCK, &rlim_new)) {
+		fprintf(stderr, "Failed to increase RLIMIT_MEMLOCK limit!\n");
+		exit(1);
+	}
+}
+
+static void handle_event(void *ctx, int cpu, void *data, __u32 data_sz)
+{
+	struct data_t *e = (struct data_t *)data;
+	struct tm *tm = NULL;
+	char ts[16];
+	time_t t;
+	if (e->ts > env.time) {
+		time(&t);
+		tm = localtime(&t);
+		strftime(ts, sizeof(ts), "%H:%M:%S", tm);
+		printf("time:%s, cpu:%d, pid:%d, gc time(us):%llu\n", ts, e->cpu,
+				e->pid, e->ts/1000);
+	}
+}
+
+static void handle_lost_events(void *ctx, int cpu, __u64 data_sz)
+{
+	printf("lost data\n");
+}
+
+static void sig_handler(int sig)
+{
+	env.exiting = true;
+}
+
+static int get_binary_path(char *path)
+{
+	char buf[64] = {0};
+	FILE *f = NULL;
+
+	snprintf(buf, 64, "cat /proc/%d/maps | grep libjvm.so | awk '{print $6}' | uniq", env.pid);
+	f = popen(buf, "r");
+	if (!f)
+		return -1;
+	fscanf(f, "%s", path);
+	pclose(f);
+
+	return 0;
+}
+
+int main(int argc, char **argv)
+{
+	static const struct argp argp = {
+		.options = opts,
+		.parser = parse_arg,
+		.doc = argp_program_doc,
+	};
+	char binary_path[BINARY_PATH_SIZE] = {0};
+	struct javagc_bpf *skel = NULL;
+	int err, i;
+	struct perf_buffer *pb = NULL;
+	struct bpf_link *links[4] = {};
+
+	err = argp_parse(&argp, argc, argv, 0, NULL, NULL);
+	if (err)
+		return err;
+
+	err = get_binary_path(binary_path);
+	if (err)
+		return err;
+
+	bump_memlock_rlimit();
+	libbpf_set_print(libbpf_print_fn);
+
+	skel = javagc_bpf__open();
+	if (!skel) {
+		fprintf(stderr, "Failed to open BPF skeleton\n");
+		return 1;
+	}
+	err = javagc_bpf__load(skel);
+	if (err) {
+		fprintf(stderr, "Failed to load and verify BPF skeleton\n");
+		goto cleanup;
+	}
+	links[0] = bpf_program__attach_usdt(skel->progs.handle_gc_start, env.pid,
+					binary_path, "hotspot", "mem__pool__gc__begin", NULL);
+	err = libbpf_get_error(links[0]);
+	if (err) {
+		fprintf(stderr, "attach usdt mem__pool__gc__begin failed:%s\n",
+				strerror(-err));
+		goto cleanup;
+	}
+	links[1] = bpf_program__attach_usdt(skel->progs.handle_gc_end, env.pid,
+					binary_path, "hotspot", "mem__pool__gc__end", NULL);
+
+	err = libbpf_get_error(links[1]);
+	if (err) {
+		fprintf(stderr, "attach usdt mem__pool__gc__end failed:%s\n",
+				strerror(-err));
+		goto cleanup;
+	}
+	links[2] = bpf_program__attach_usdt(skel->progs.handle_gc_start, env.pid,
+					binary_path, "hotspot", "gc__begin", NULL);
+	err = libbpf_get_error(links[2]);
+	if (err) {
+		fprintf(stderr, "attach usdt gc__begin failed:%s\n", strerror(-err));
+		goto cleanup;
+	}
+	links[3] = bpf_program__attach_usdt(skel->progs.handle_gc_end, env.pid,
+				binary_path, "hotspot", "gc__end", NULL);
+	err = libbpf_get_error(links[3]);
```

Comment:
With libbpf strict mode, you don't need `libbpf_get_error`, just check against `NULL`.

Reply:
Fix it